File: LeetCode/top-interview-questions-medium/LinkedList/add-two-numbers.py

#  Definition for singly-linked list.
#  class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        c = 0
        ans = ListNode()
        while l1 or l2:
            s = c
            s += l1.val if l1 else 0
            s += l2.val if l2 else 0
            c = 1 if s >= 10 else 0
            s = s % 10 if s >= 10 else s
            ans.next = ListNode(val=s)
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        if c > 0:
            ans.next = ListNode(val=c)
        return ans.next

File: LeetCode/tags/linkedlist/add-two-numbers.cpp
#include<bits/stdc++.h>
using namespace std;
#define endl '\n'
#define MOD 1000000007
typedef vector<int> vi;
typedef vector<vector<int>> vii;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry=0, curr=0; ListNode dummy(-1), *node=&dummy;
        while(l1 || l2 || carry)  {
            curr = carry;
            if(l1) curr += l1->val, l1=l1->next;
            if(l2) curr += l2->val, l2=l2->next;
            carry = curr / 10;
            curr = curr % 10;
            node->next= new ListNode(curr);
            node = node->next;
        }
        return dummy.next;
    }
};


